scrapy/commands/genspider.py

```diff
@@ -69,7 +69,18 @@ def run(self, args, opts):
         try:
             spidercls = self.crawler_process.spider_loader.load(name)
         except KeyError:
-            pass
+            if not self.settings.get('NEWSPIDER_MODULE'):
+                if exists(name + ".py"):
+                    print("File %r already exists in the current directory" % (name + ".py"))
+                    return
+            else:
```

Comment:
Could you add a test to cover the case where `NEWSPIDER_MODULE` is defined?

Reply:
Do you mean replacing the `else` with an `elif self.settings.get('NEWSPIDER_MODULE)`?, The `else` block should handle that case right?

Reply:
I've added a test to the case @elacuesta mentioned, `test_same_filename_as_existing_spider` under `GenspiderCommandTest` where `NEWSPIDER_MODULE` will be defined when the test is run. Could you please elaborate? don't quite understand what you mean.

Reply:
> where NEWSPIDER_MODULE will be defined when the test is run

Where is it defined? Or do you mean that people running the tests must define it manually somewhere before running the test?

Reply:
Sorry, should have phrased it better. From what I understand of the code base, when the `scrapy` command is run in the command line, the `settings` object is set in line 112 of the [cmdline](https://github.com/scrapy/scrapy/blob/master/scrapy/cmdline.py#L112) module, digging further into `get_project_settings()`, I found that the value of `NEWSPIDER_MODULE` **is set here to either an empty string or the module where the spiders are located** in case of a project. This `settings` object is then passed on to [cmd](https://github.com/scrapy/scrapy/blob/master/scrapy/cmdline.py#L137) in line 137.

At this point `cmd` **is already set to** `genspider` and the line 137 `cmd.settings = settings` assigns the previously mentioned `settings` object to `genspider.settings`. This settings dictionary is what I'm accessing through `self.settings`. So essentially the `NEWSPIDER_MODULE` property will always be defined (either to `''` or the spiders module)

Reply:
Makes sense.

So, given your code also handles the scenario where the `NEWSPIDER_MODULE` setting is defined with a non-empty string as value, I think we need a test for that scenario.

Reply:
The value of `NEWSPIDER_MODULE` is set implicitly when the `scrapy` command is run, I think the only way we can test it for both the scenarios is if we test running `genspider` as a standalone command and then within a project scope, which I think I have handled in the tests file.

Reply:
`test_same_filename_as_existing_spider` for when the command is run within a project and `test_same_name_as_existing_file` for when the command is run as a standalone

Reply:
OK, I finally understand [why the tests cover `NEWSPIDER_MODULE` usage](https://github.com/scrapy/scrapy/blob/master/scrapy/templates/project/module/settings.py.tmpl#L13). Sorry! :flushed: